### PR TITLE
DAL-2874: Add DalContext

### LINKED ISSUE #2874: Add `DalContext` to dal crate

### LINKED ISSUE #2875: Remove the metrics from dal crate

### PR DESCRIPTION
**Pull request by BohuTANG**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Summary about this PR:
1. Introduce `DalContext` for binding metrics and add DalContext to interceptors
2. Replace `TenantLabel` with `DalContext`
3. Rename `common/dal/src/impls` -> `common/dal/src/accessors`
4. Rename `common/dal/src/metrics` -> `common/dal/src/interceptors`
5. Reorder `query/src/sessions/context.rs` method orders

## Changelog

- Improvement

## Related Issues

Fixes #2874 
Fixes #2875

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/dal/src/context.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /// Increment write bytes.
+    pub fn inc_write_bytes(&self, bytes: usize) {
+        let mut metrics = self.metrics.write();
+        metrics.write_bytes += bytes;
+    }
+
+    pub fn get_metrics(&self) -> DalMetrics {
+        self.metrics.read().clone()
```
Comment (Veeupup) [inline common/dal/src/context.rs:50]: it seems that `DalMetrics` is a POD(with only two members), so maybe `Copy` semantic is enough?

Comment (BohuTANG) [reply common/dal/src/context.rs:50]: Metrics in the DalContext is ` metrics: Arc<RwLock<DalMetrics>>,` clone is easy to do that

Comment (Veeupup) [reply common/dal/src/context.rs:50]: got it.

**Comment by drmingdrmer** [comment]
/lgtm

Appreciating these clean concepts and clean impls!

**Pull request merged**